src/solvers/lowering/byte_operators.cpp

```diff
@@ -19,6 +19,168 @@ Author: Daniel Kroening, kroening@kroening.com
 
 #include "flatten_byte_extract_exceptions.h"
 
+static exprt bv_to_expr(
+  const exprt &bitvector_expr,
+  const typet &target_type,
+  const namespacet &ns)
+{
+  if(
+    can_cast_type<bitvector_typet>(target_type) ||
+    target_type.id() == ID_c_enum || target_type.id() == ID_c_enum_tag)
+  {
+    return simplify_expr(
+      typecast_exprt::conditional_cast(bitvector_expr, target_type), ns);
+  }
+
+  if(target_type.id() == ID_struct || target_type.id() == ID_struct_tag)
+  {
+    const struct_typet &struct_type = to_struct_type(ns.follow(target_type));
+    const struct_typet::componentst &components = struct_type.components();
+
+    exprt::operandst operands;
+    operands.reserve(components.size());
+    std::size_t offset = 0;
+    for(const auto &comp : components)
+    {
+      auto component_bits = pointer_offset_bits(comp.type(), ns);
+      std::size_t bits;
+      if(component_bits.has_value())
+        bits = numeric_cast_v<std::size_t>(*component_bits);
+      else
+        bits = to_bitvector_type(bitvector_expr.type()).get_width() - offset;
+
+      if(bits == 0)
+      {
+        operands.push_back(constant_exprt(irep_idt(), comp.type()));
+        continue;
+      }
+
+      bitvector_typet type(bitvector_expr.type().id(), bits);
+      operands.push_back(bv_to_expr(
+        extractbits_exprt(bitvector_expr, offset + bits - 1, offset, type),
+        comp.type(),
+        ns));
+
+      if(component_bits.has_value())
+        offset += bits;
+    }
+
+    return struct_exprt(std::move(operands), target_type);
+  }
+  else if(target_type.id() == ID_array)
+  {
+    const array_typet &array_type = to_array_type(target_type);
+    auto num_elements = numeric_cast<std::size_t>(array_type.size());
+    auto subtype_bits = pointer_offset_bits(array_type.subtype(), ns);
+
+    const std::size_t total_width =
+      to_bitvector_type(bitvector_expr.type()).get_width();
+    if(!num_elements.has_value())
+    {
+      if(!subtype_bits.has_value() || *subtype_bits == 0)
+        num_elements = total_width;
+      else
+        num_elements = total_width / numeric_cast_v<std::size_t>(*subtype_bits);
+    }
+
+    exprt::operandst operands;
+    operands.reserve(*num_elements);
+    for(std::size_t i = 0; i < *num_elements; ++i)
+    {
+      if(subtype_bits.has_value())
+      {
+        const std::size_t subtype_bits_int =
+          numeric_cast_v<std::size_t>(*subtype_bits);
+        bitvector_typet type(bitvector_expr.type().id(), subtype_bits_int);
+        operands.push_back(bv_to_expr(
+          extractbits_exprt(
+            bitvector_expr,
+            (i + 1) * subtype_bits_int - 1,
```

Comment:
Suggest parens to clarify intent `((i + 1) * subtype_bits_int) - 1`

Reply:
Done.